phac-nml/irida-next
# GraphQL create direct upload

## What does this PR do and why?

Adds CreateDirectUpload mutation to the GraphQL API
Fixes #390 

## Screenshots or screen recordings

N/A

## How to set up and validate locally

1. setup branch
```bash
bundle
./scripts/clean_dev.sh
```

2. run irida next
3. sign in with `admin@email.com`
4. go to `http://localhost:3000/graphiql` to access graphql interface
5. execute a mutaion to create a direct upload blob

```graphql
mutation makeblob {
  createDirectUpload(input: {
    filename: "dev.to", # file name
    contentType: "image/jpeg", # file content type
    checksum: "Z3Yzc2Q5iA5eXIgeTJn", # checksum
    byteSize: 2019 # size in bytes
  })
  {
    directUpload{
      # url,
      headers,
      blobId,
      signedBlobId
    }
  }
}
```

6. verify appropriate data is returned. Example:

```json
{
  "data": {
    "createDirectUpload": {
      "directUpload": {
        "url": "http://localhost:3000/rails/active_storage/disk/eyJfcmFpbHMiOnsiZGF0YSI6eyJrZXkiOiJxN253dzFpZzN3bHV3MTR5djJteDNwMzM1eGdsIiwiY29udGVudF90eXBlIjoiaW1hZ2UvanBlZyIsImNvbnRlbnRfbGVuZ3RoIjoxMjMsImNoZWNrc3VtIjoiWjNZemMyUTVpQTVlWElnZVRKbiIsInNlcnZpY2VfbmFtZSI6ImxvY2FsIn0sImV4cCI6IjIwMjQtMDEtMjlUMTg6Mjg6MTUuNjU2WiIsInB1ciI6ImJsb2JfdG9rZW4ifX0=--333580b8f7a36664e97d295a0330071fc50c7801",
        "headers": "{\"Content-Type\":\"image/jpeg\"}",
        "blobId": "76",
        "signedBlobId": "eyJfcmFpbHMiOnsiZGF0YSI6NzYsInB1ciI6ImJsb2JfaWQifX0=--532c0032507c900f79c773f31616c2852588951e"
      }
    }
  }
}
```


## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
